core/domain/state_domain.py

```diff
@@ -353,192 +406,1900 @@ def __init__(
     def to_dict(self):
         """Returns a dict representing this Solution domain object.
 
-        Returns:
-            dict. A dict mapping all fields of Solution instance.
-        """
-        return {
-            'answer_is_exclusive': self.answer_is_exclusive,
-            'correct_answer': self.correct_answer,
-            'explanation': self.explanation.to_dict(),
-        }
+        Returns:
+            dict. A dict mapping all fields of Solution instance.
+        """
+        return {
+            'answer_is_exclusive': self.answer_is_exclusive,
+            'correct_answer': self.correct_answer,
+            'explanation': self.explanation.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, interaction_id, solution_dict):
+        """Return a Solution domain object from a dict.
+
+        Args:
+            interaction_id: str. The interaction id.
+            solution_dict: dict. The dict representation of Solution object.
+
+        Returns:
+            Solution. The corresponding Solution domain object.
+        """
+        explanation = SubtitledHtml.from_dict(solution_dict['explanation'])
+        explanation.validate()
+        return cls(
+            interaction_id,
+            solution_dict['answer_is_exclusive'],
+            interaction_registry.Registry.get_interaction_by_id(
+                interaction_id).normalize_answer(
+                    solution_dict['correct_answer']),
+            explanation)
+
+    def to_proto(self):
+        """Returns a proto representation of the solution object.
+
+        Returns:
+            BaseSolution. The BaseSolution proto object.
+        """
+        base_solution_proto = state_pb2.BaseSolution(
```

Comment:
What is the difference between a solution and a "base solution"?

Reply:
As per proto, the BaseSolution is:

```proto
message BaseSolution {
  option (structure_proto_version_type) = STATE_PROTO_VERSION;

  // The explanation of the solution to show the user.
  SubtitledText explanation = 1;
}
```

and `message Solution {}` is different for every interaction. The `BaseSolution` contains the explanation part only.